File: Group9_PieChart/QUESTION B/colored_pie_chart.cpp
#include <GL/glut.h>
#include <cmath>
#include <cstdio>
#include <string>

// Window dimensions
const int WINDOW_WIDTH = 800;
const int WINDOW_HEIGHT = 800;

// Pie chart data values and labels
static float values[] = {36.0f, 41.0f, 19.0f, 28.0f, 30.0f, 16.0f};
static const char *labels[] = {"Avocado", "Orange", "Banana", "Kiwifruit", "Mangos", "Grapes"};

// RGB color values for each fruit slice
static float colors[][3] = {
    {0.82f, 0.81f, 0.41f},  // Avocado
    {0.93f, 0.55f, 0.14f},  // Orange
    {1.0f, 0.87f, 0.35f},   // Banana
    {0.43f, 0.51f, 0.04f},  // Kiwifruit
    {1.0f, 0.51f, 0.26f},   // Mango
    {0.44f, 0.18f, 0.66f}   // Grapes
};

// Number of slices
static const int NUM_SLICES = sizeof(values) / sizeof(values[0]);

// Chart title
static const char *chartTitle = "Youth Fruit Preferences in Gachororo";

// Pi constant for angle conversion
const float PI = 3.14159265f;

// Utility function to draw text on screen using bitmap fonts
void drawBitmapText(float x, float y, const char *string) {
    glRasterPos2f(x, y); // Set text position
    for (const char *c = string; *c != '\0'; c++) {
        glutBitmapCharacter(GLUT_BITMAP_HELVETICA_12, *c); // Render each character
    }
}

// Main display callback
void display() {
    glClear(GL_COLOR_BUFFER_BIT);  // Clear background
    glLoadIdentity();

    // Calculate total value for percentage calculations
    float total = 0.0f;
    for (int i = 0; i < NUM_SLICES; ++i) {
        total += values[i];
    }

    // Pie chart parameters
    float centerX = 0.0f, centerY = 0.0f, radius = 0.6f;
    float currentAngle = 0.0f;

    // Draw chart title
    glColor3f(0.0f, 0.0f, 0.0f);
    drawBitmapText(-0.4f, 0.9f, chartTitle);

    // Draw each pie slice
    for (int i = 0; i < NUM_SLICES; ++i) {
        float sliceAngle = 360.0f * values[i] / total;  // Convert value to angle in degrees

        // Draw filled colored slice using triangle fan
        glColor3f(colors[i][0], colors[i][1], colors[i][2]);
        glBegin(GL_TRIANGLE_FAN);
        glVertex2f(centerX, centerY); // Pie center
        for (float angle = currentAngle; angle <= currentAngle + sliceAngle; angle += 1.0f) {
            float rad = angle * PI / 180.0f;
            glVertex2f(centerX + cos(rad) * radius, centerY + sin(rad) * radius);
        }
        // Close the fan
        float endRad = (currentAngle + sliceAngle) * PI / 180.0f;
        glVertex2f(centerX + cos(endRad) * radius, centerY + sin(endRad) * radius);
        glEnd();

        // Optional: draw black outline around each slice
        glColor3f(0.0f, 0.0f, 0.0f);
        glLineWidth(1.0f);
        glBegin(GL_LINE_LOOP);
        glVertex2f(centerX, centerY);
        for (float angle = currentAngle; angle <= currentAngle + sliceAngle; angle += 1.0f) {
            float rad = angle * PI / 180.0f;
            glVertex2f(centerX + cos(rad) * radius, centerY + sin(rad) * radius);
        }
        glEnd();

        // Compute label position at the middle angle of the slice
        float midAngle = currentAngle + sliceAngle / 2.0f;
        float midRad = midAngle * PI / 180.0f;
        float labelRadius = radius + 0.15f;

        // Fine-tune label positions for better spacing
        if (i == 2) labelRadius += 0.1f;  // Banana
        if (i == 3) labelRadius += 0.05f; // Kiwifruit

        float labelX = centerX + cos(midRad) * labelRadius;
        float labelY = centerY + sin(midRad) * labelRadius;

        // Format label with percentage
        char labelText[50];
        sprintf(labelText, "%s (%.1f%%)", labels[i], 100.0f * values[i] / total);

        // Draw label
        glColor3f(0.0f, 0.0f, 0.0f);
        drawBitmapText(labelX - 0.05f, labelY - 0.02f, labelText);

        // Update angle for next slice
        currentAngle += sliceAngle;
    }

    glutSwapBuffers();  // Swap buffers for smooth double buffering
}

// Reshape callback to handle window resizing
void reshape(int w, int h) {
    glViewport(0, 0, w, h);              // Reset viewport
    glMatrixMode(GL_PROJECTION);        // Switch to projection matrix
    glLoadIdentity();                   
    gluOrtho2D(-1.0, 1.0, -1.0, 1.0);    // Set 2D orthographic projection
    glMatrixMode(GL_MODELVIEW);         // Switch back to modelview matrix
}

// Program entry point
int main(int argc, char **argv) {
    glutInit(&argc, argv);                                      // Initialize GLUT
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB);                // Double buffer and RGB color
    glutInitWindowSize(WINDOW_WIDTH, WINDOW_HEIGHT);            // Set window size
    glutCreateWindow("Fruit Preferences Pie Chart");            // Create window
    glClearColor(1.0f, 1.0f, 1.0f, 1.0f);                        // Set white background
    glutDisplayFunc(display);                                   // Register display callback
    glutReshapeFunc(reshape);                                   // Register reshape callback
    glutMainLoop();                                             // Enter main event loop
    return 0;
}

File: Group9_PieChart/QUESTION B/colored_pie_chart.py
import numpy as np
from OpenGL.GL import *
from OpenGL.GLU import *
from OpenGL.GLUT import *
import math

# Survey data
fruits = ['Avocado', 'Orange', 'Banana', 'Kiwifruit', 'Mango', 'Grapes']
people = [36, 41, 19, 28, 30, 16]
total = sum(people)  # Total number of responses
percentages = [round((count / total) * 100, 1) for count in people]  # Percentage per fruit

# RGB colors for each fruit slice
fruit_colors = [
    (0.82, 0.81, 0.41),   # Avocado 
    (0.93, 0.55, 0.14),   # Orange
    (1.0, 0.87, 0.35),    # Banana 
    (0.43, 0.51, 0.04),   # Kiwifruit 
    (1.0, 0.51, 0.26),    # Mango 
    (0.44, 0.18, 0.66)    # Grapes 
]

def draw_pie_chart():
    glClear(GL_COLOR_BUFFER_BIT)  # Clear the screen
    glLoadIdentity()              # Reset transformations

    # Draw the chart title
    glColor3f(0, 0, 0)  # Black color for text
    glRasterPos2f(-0.9, 0.9)  # Position the title near the top-left
    for c in "Youth Fruit Preferences in Gachororo":
        glutBitmapCharacter(GLUT_BITMAP_HELVETICA_18, ord(c))

    radius = 0.6          # Radius of the pie chart
    angle_start = 0       # Initial angle (in radians)

    # Loop through each fruit category to draw a slice
    for i in range(len(people)):
        # Calculate the angle that this slice will span
        angle_end = angle_start + 2 * math.pi * people[i] / total

        # Begin drawing a filled fan shape (slice)
        glBegin(GL_TRIANGLE_FAN)
        glColor3f(*fruit_colors[i])  # Set the slice color
        glVertex2f(0, 0)  # Center of the pie

        # Draw the edge of the slice from start to end angle
        segments = 50  # Controls smoothness of curve
        for j in range(segments + 1):
            angle = angle_start + (angle_end - angle_start) * j / segments
            glVertex2f(radius * math.cos(angle), radius * math.sin(angle))
        glEnd()

        # Position label near the outer edge of the slice
        mid_angle = (angle_start + angle_end) / 2
        label_x = radius * math.cos(mid_angle)
        label_y = radius * math.sin(mid_angle)

        # Slightly offset the label outward
        offset = 0.12
        label_x += offset * math.cos(mid_angle)
        label_y += offset * math.sin(mid_angle)

        # Draw the label (fruit name + percentage)
        glColor3f(0, 0, 0)  # Label text color
        glRasterPos2f(label_x, label_y)
        label = f"{fruits[i]} ({percentages[i]}%)"
        for c in label:
            glutBitmapCharacter(GLUT_BITMAP_9_BY_15, ord(c))

        angle_start = angle_end  # Update for next slice

    glutSwapBuffers()  # Swap the double buffers (display result)

def init():
    # Set background color to white
    glClearColor(1.0, 1.0, 1.0, 1.0)

    # Set the 2D orthographic projection (no perspective)
    gluOrtho2D(-1.0, 1.0, -1.0, 1.0)

def main():
    glutInit()  # Initialize GLUT
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB)  # Double buffering + RGB color mode
    glutInitWindowSize(800, 800)  # Window size in pixels
    glutCreateWindow(b"Fruit Preferences Pie Chart")  # Create window with title
    glutDisplayFunc(draw_pie_chart)  # Set the display callback
    init()  # Custom initialization
    glutMainLoop()  # Enter the event processing loop

if __name__ == "__main__":
    main()
